contrib/format-esri/src/main/java/org/apache/drill/exec/store/esri/ShpBatchReader.java

```diff
@@ -0,0 +1,334 @@
+
+    while( !rowWriter.isFull()){
+      Object[] dbfRow = dbfReader.nextRecord();
+      geom = geomCursor.next();
+      if(geom == null) {
+        return false;
+      }
+      processShapefileSet(rowWriter, geomCursor.getGeometryID(), geom, dbfRow);
+    }
+    return true;
+  }
+
+  private void openFile(FileSchemaNegotiator negotiator) {
+    try {
+      fileReaderShp = negotiator.fileSystem().open(split.getPath());
+      byte[] shpBuf = new byte[fileReaderShp.available()];
+      fileReaderShp.readFully(shpBuf);
+
+      ByteBuffer byteBuffer = ByteBuffer.wrap(shpBuf);
+      byteBuffer.position(byteBuffer.position() + 100);
+
+      ShapefileReader shpReader = new ShapefileReader();
+      geomCursor = shpReader.getGeometryCursor(byteBuffer);
+
+      fileReaderDbf = negotiator.fileSystem().open(hadoopDbf);
+      dbfReader = new DbfReader(fileReaderDbf);
+
+      fileReaderPrj = negotiator.fileSystem().open(hadoopPrj);
+      byte[] prjBuf = new byte[fileReaderPrj.available()];
+      fileReaderPrj.readFully(prjBuf);
+      fileReaderPrj.close();
+
+      String wktReference = new String(prjBuf);
+
+      String sridPatternText = "AUTHORITY\\[\"\\w+\"\\s*,\\s*\"*(\\d+)\"*\\]\\]$";
+      Pattern pattern = Pattern.compile(sridPatternText);
+      Matcher matcher = pattern.matcher(wktReference);
+      if (matcher.find()) {
+        this.srid = Integer.parseInt(matcher.group(1));
+        this.spatialReference = SpatialReference.create(srid);
+      }
+
+      logger.debug("Processing Shape File : {}", hadoopShp);
+    } catch (IOException e) {
+      throw UserException
+        .dataReadError(e)
+        .message("Failed to open open input file: %s", split.getPath())
+        .addContext("User name", negotiator.userName())
+        .build(logger);
+    }
+  }
+
+  private String byteArrayToString(byte[] in) {
+    char out[] = new char[in.length * 2];
+    for (int i = 0; i < in.length; i++) {
+      out[i * 2] = "0123456789ABCDEF".charAt((in[i] >> 4) & 15);
+      out[i * 2 + 1] = "0123456789ABCDEF".charAt(in[i] & 15);
+    }
+    return new String(out);
+  }
+
+  private void processShapefileSet(RowSetLoader rowWriter, final int gid, final Geometry geom, final Object[] dbfRow) {
+    rowWriter.start();
+
+    gidWriter.setInt(gid);
+    sridWriter.setInt(srid);
+    shapeTypeWriter.setString(geom.getType().toString());
+    final ByteBuffer buf = (ByteBuffer) OGCGeometry.createFromEsriGeometry(geom, spatialReference).asBinary();
+    final byte[] bytes = buf.array();
+    geomWriter.setBytes(bytes, bytes.length);
+    writeDbfRow(dbfRow, rowWriter);
+
+    rowWriter.save();
+  }
+
+  private void writeDbfRow(final Object[] dbfRow, RowSetLoader rowWriter) {
+    int dbfFieldCount = dbfReader.getHeader().getFieldsCount();
+
+    for (int i = 0; i < dbfFieldCount; i++) {
+      DbfField field = dbfReader.getHeader().getField(i);
+
+      if (dbfRow[i] == null) {
+        continue;
+      }
+
+      switch (field.getDataType()) {
```

Comment:
As suggested in another plugin, if fields are in a fixed order, would be faster to create classes for each type, then dispatch to the correct function. Index into an array for the column object. See example in that other review.